### DISCUSSION TITLE
Different results for code with __future__.annotations and without

### DISCUSSION THREAD

**Discussion by InSyncWithFoo** [microsoft/pyright#7394] [General]
The following code is originally from [this StackOverflow question](https://stackoverflow.com/q/78101227).

[With `__future__.annotations`](https://pyright-play.net/?pythonVersion=3.12&strict=true&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwFgAoFgYwBtj76oBxdGhqkA5vQBcLKNKgATNMCgcs9JAAp6aDsACUUALQA%2BJSqQBtVSAC64qADoHUmfMUALLRzAatwADRRSYgg0CRNVCyRrPSMoADk6NFsHOxYWdAA3NGIOGCQAT1w0NQE0ISxRejUdO2VVKr0oAGIw80srFiA):

```py
from __future__ import annotations

class Greetings:
    def list(self) -> list[str]: ...
    def hello(self, names: list[str]) -> None: ...

reveal_type(Greetings().list())  # list[str]
```

[Without](https://pyright-play.net/?pythonVersion=3.12&strict=true&code=MYGwhgzhAEDiBOBTRAXAlgOwOYQFwFgAoaE6AE0QDNoQ0IUAKCRESgSmgFoA%2BGulANr14AXVzQAdFKKlyVaAAsWIAPZMWlADTQMYALaI8fekJSiOPaADkVGROKkSiRJADdEYEAH0UATwAOiAwIyOjYEAxsErT0kRzQAMTGgsIiREA):

```py
class Greetings:
    def list(self) -> list[str]: ...
    def hello(self, names: list[str]) -> None: ...
    #                      ~~~~ => Expected type expression but received "(self: Self@Greetings) -> list[str]"
    #               ~~~~~ => Type of parameter "names" is unknown

reveal_type(Greetings().list())  # list[str]
```

Note how the `reveal_type()` results are correct in both examples (`Greetings().list()[0:]` is correctly inferred as `list[str]`, for instance).

Is this intended?

**Comment by erictraut** [comment]
Yes, this is intended. When you're not using `from __future__ import annotations`, the second code snippet will generate a runtime exception, and pyright is correct in flagging it as an error.

**Comment by InSyncWithFoo** [comment]
Ignore me... I must have been too sleepy when I posted this `:P`
